feat(regex_parser): Implement `RegExp` parser

leaysgur opened the pull request:
Part of #1164

## Progress updates 🗞️ 

Waiting for the review and advice, while thinking how to handle escaped string when `new RegExp(pat)`.

## TODOs

- [x] `RegExp(Literal = Body + Flags)#parse()` structure
- [x] Base `Reader` impl to handle both unicode(u32) and utf-16(u16) units
- [x] Global `Span` and local offset conversion
- [x] Design AST shapes
  - [x] Keep `enum` size small by `Box<'a, T>`
  - [x] Rework AST shapes 
- [x] Split body and flags w/ validating literal
- [x] Parse `RegExpFlags`
- [x] Parse `RegExpBody` = `Pattern`
- [x] Parse `Pattern` > `Disjunction`
- [x] Parse `Disjunction` > `Alternative`
- [x] Parse `Alternative` > `Term`
- [x] Parse `Term` > `Assertion`
	- [x] Parse `BoundaryAssertion`
	- [x] Parse `LookaroundAssertion`
- [x] Parse `Term` > `Quantifier`
- [x] Parse `Term` > `Atom`
	- [x] Parse `Atom` > `PatternCharacter`
	- [x] Parse `Atom` > `.`
	- [x] Parse `Atom` > `\AtomEscape`
		- [x] Parse `\AtomEscape` > `DecimalEscape`
		- [x] Parse `\AtomEscape` > `CharacterClassEscape`
			- [x] Parse `CharacterClassEscape` > `\d, \D, \s, \S, \w, \W`
			- [x] Parse `CharacterClassEscape` > `\p{UnicodePropertyValueExpression}, \P{UnicodePropertyValueExpression}`
		- [x] Parse `\AtomEscape` > `CharacterEscape`
			- [x] Parse `CharacterEscape` > `ControlEscape`
			- [x] Parse `CharacterEscape` > `c AsciiLetter`
			- [x] Parse `CharacterEscape` > `0`
			- [x] Parse `CharacterEscape` > `HexEscapeSequence`
			- [x] Parse `CharacterEscape` > `RegExpUnicodeEscapeSequence`
			- [x] Parse `CharacterEscape` > `IdentityEscape`
		- [x] Parse `\AtomEscape` > `kGroupName`
	- [x] Parse `Atom` > `[CharacterClass]`
    	- [x] Parse `[CharacterClass]` > `ClassContents` > `[~UnicodeSetsMode] NonemptyClassRanges`
    	- [x] Parse `[CharacterClass]` > `ClassContents` > `[+UnicodeSetsMode] ClassSetExpression`
          - [x] Parse `ClassSetExpression` > `ClassUnion`
          - [x] Parse `ClassSetExpression` > `ClassIntersection`
          - [x] Parse `ClassSetExpression` > `ClassSubtraction`
          - [x] Parse `ClassSetExpression` > `ClassSetOperand`
          - [x] Parse `ClassSetExpression` > `ClassSetRange`
          - [x] Parse `ClassSetExpression` > `ClassSetCharacter`
	- [x] Parse `Atom` > `(GroupSpecifier)`
	- [x] Parse `Atom` > `(?:Disjunction)`
- [x] Annex B
    - [x] Parse `QuantifiableAssertion`
	- [x] Parse `ExtendedAtom`
      - [x] Parse `ExtendedAtom` > `\ [lookahead = c]`
      - [x] Parse `ExtendedAtom` > `InvalidBracedQuantifier`
      - [x] Parse `ExtendedAtom` > `ExtendedPatternCharacter`
      - [x] Parse `ExtendedAtom` > `\AtomEscape` > `CharacterEscape` > `LegacyOctalEscapeSequence`
- [x] Early errors
	- [x] Pattern :: Disjunction(1/2)
	- [x] Pattern :: Disjunction(2/2)
	- [x] QuantifierPrefix :: { DecimalDigits , DecimalDigits }
	- [x] ExtendedAtom :: InvalidBracedQuantifier (Annex B)
	- [x] AtomEscape :: k GroupName
	- [x] AtomEscape :: DecimalEscape
	- [x] NonemptyClassRanges :: ClassAtom - ClassAtom ClassContents(1/2)
	- [x] NonemptyClassRanges :: ClassAtom - ClassAtom ClassContents(2/2)
	- [x] NonemptyClassRanges :: ClassAtom - ClassAtom ClassContents(Annex B)
	- [x] NonemptyClassRangesNoDash :: ClassAtomNoDash - ClassAtom ClassContents(1/2)
	- [x] NonemptyClassRangesNoDash :: ClassAtomNoDash - ClassAtom ClassContents(2/2)
	- [x] NonemptyClassRangesNoDash :: ClassAtomNoDash - ClassAtom ClassContents(Annex B)
	- [x] RegExpIdentifierStart :: \ RegExpUnicodeEscapeSequence
	- [x] RegExpIdentifierStart :: UnicodeLeadSurrogate UnicodeTrailSurrogate
	- [x] RegExpIdentifierPart :: \ RegExpUnicodeEscapeSequence
	- [x] RegExpIdentifierPart :: UnicodeLeadSurrogate UnicodeTrailSurrogate
	- [x] UnicodePropertyValueExpression :: UnicodePropertyName = UnicodePropertyValue(1/2)
	- [x] UnicodePropertyValueExpression :: UnicodePropertyName = UnicodePropertyValue(2/2)
	- [x] UnicodePropertyValueExpression :: LoneUnicodePropertyNameOrValue(1/2)
	- [x] UnicodePropertyValueExpression :: LoneUnicodePropertyNameOrValue(2/2)
	- [x] CharacterClassEscape :: P{ UnicodePropertyValueExpression }
	- [x] CharacterClass :: [^ ClassContents ]
	- [x] NestedClass :: [^ ClassContents ]
	- [x] ClassSetRange :: ClassSetCharacter - ClassSetCharacter
- [x] Add `Span` to `Err(OxcDiagnostic::error())` calls
- [x] Perf improvement 
	- [x] `Reader#peek()` should avoid `iter.next()` equivalent
	- [x] ~~Use `char` everywhere and split and push 2 surrogates(pair) for `Character`?~~
	- [x] ~~Try 1(+1) loop parsing for capturing groups?~~

## Follow up

- [x] @Boshen Test suite > #4242
  - [x] Investigate CI errors...
- Next...
  - Support ES2025 Duplicate named capturing groups?
  - Support ES20XX Stage3 Modifiers?

Boshen commented on crates/oxc_regexp_parser/_oxc_js_regex/ast_builder.rs:
```diff
@@ -0,0 +1,73 @@
+        String::from_str_in(value, self.allocator).into_bump_str()
+    }
+
+    pub fn copy<T>(&self, src: &T) -> T {
```
Let's remove this API, it's totally unsafe :-)

Boshen commented on crates/oxc_regexp_parser/_oxc_js_regex/ast_kind.rs:
We won't be needing this, can be removed. We'll use a standard visitor pattern.

Boshen commented on crates/oxc_regexp_parser/src/ast.rs:
For some of the `enum`s, we need to box them

https://github.com/oxc-project/oxc/blob/445603444f98402fb66a4a956b1023611bc66133/crates/oxc_ast/src/lib.rs#L34-L75

leaysgur replied:
Even in the current code, Box has already been applied to the part of the code where the recursion warning was issued.

Perhaps this comment meant that all `enum` members should be `Box`ed?

Or is it just guidance on where `Box` should be used and the code seems fine as it is?

Boshen replied:
With rust-analyzer, you can "see" the enum size (I don't know the hotkey in vscode):

you can see `size = 16` here, it should be boxed if it's allocated a lot and the size is big.

<img width="887" alt="image" src="https://github.com/oxc-project/oxc/assets/1430279/a09f85de-fc48-4d0b-93d5-d89939934e35">

leaysgur replied:
Oh~, I got it! thank you~ 😃 

|before|after|
|------|----|
|![image](https://github.com/oxc-project/oxc/assets/6259812/a250905e-405a-4ee9-8776-30b4453c7d67)|![image](https://github.com/oxc-project/oxc/assets/6259812/904954a1-cb6a-40c8-ae72-5e8d2fa5f0ac)|

I'll keep every `enum` size to be 16.

Boshen commented on crates/oxc_regexp_parser/src/ast_builder.rs:
This builder pattern is unnecessary, can be removed.

Boshen commented on crates/oxc_regexp_parser/src/parser/literal_parser.rs:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    // NOTE: Should return `ParserReturn { (empty)literal, errors }`?
+    pub fn parse(self) -> Result<ast::RegExpLiteral<'a>> {
```
Are we building a recoverable parser? Can `Result` if this parser is not recoverable.

leaysgur replied:
Probably no? (I don't think I even properly understood what recoverable meant in the first place, being inexperienced...)

This comment was a remnant of the process of following the code of `oxc_parser`.

leaysgur commented:
Thanks for the quick review!

Next time, I will:

- [ ] Apply `Box` to `enum`
  - with reading that comment and link
- [ ] Remove ast_builder

leaysgur commented:
@Boshen  I have managed to parse minimal patterns such as `/ab|c/`.

I would like to confirm if this direction is correct, so could you provide a review or general advice for implementing the parser at this initial stage?

Boshen commented on crates/oxc_regexp_parser/src/parser/body_parser/reader.rs:
```diff
@@ -0,0 +1,226 @@
+    fn peek_nth(&self, n: usize) -> Option<u32> {
+        let nth = self.index + n;
+
+        // NOTE: This may affect performance?
```
Yes, `.chars().next()` is really expensive.

Take a look at `lookahead` in `crates/oxc_parser/src/lexer/mod.rs.

Boshen commented on crates/oxc_regexp_parser/src/parser/body_parser/reader.rs:
```diff
@@ -0,0 +1,226 @@
+pub struct Reader<'a> {
+    source: &'a str,
+    char_indices: std::str::CharIndices<'a>,
```
I have a feeling that we should use bytes over chars ... but I'm not sure 😅

Boshen reviewed (commented):
The architecture is really good!

You may want to consider adding "context" early so you don't get caught by surprise when you need it.

```
Pattern[UnicodeMode, UnicodeSetsMode, NamedCaptureGroups]
        ^^^^^^^^^^^  ^^^^^^^^^^^^^^^  ^^^^^^^^^^^^^^^^^^
```

Take a look at `context` in crates/oxc_parser/src/cursor.rs

leaysgur commented:
Thanks for your time!

> The architecture is really good!

I'm relieved to hear that. 😎

I'll leave a TODO for `chars.nth()` within `peek()` to be addressed later.

Boshen commented:
Now that we have some of the implementation working, we should think about how to support the regex eslint rules 🤔

Sysix commented:
@leaysgur hello, im currently working on a smaller version of regex groups, maybe u find some usefull snippets here:
https://github.com/oxc-project/oxc/pull/4096/files

interesting method:
- get_regex_group_by_open_bracet

Currently Missing implementations:
- Named Captures (named capture can be accessed by it's number reference. `(?<foo>a)(\1)` is valid)

rzvxa commented:
This is awesome, I'm looking forward to this PR😍

I always had a theory that there are only 5 people on StackOverflow who write all the regex examples and everyone else just copies them into production. If that theory is correct I bet you'd be the 6th after this😆

leaysgur commented:
That's the truth. 😅
I will do my best to reverse my position from copying to be copied side!

leaysgur commented:
### Perf trial

<details>
<summary>Current</summary>

```rs
pub struct Reader<'a> {
    source: &'a str,
    unicode_mode: bool,
    index: usize,
    // NOTE: For now, this exists only for `span_position()` method.
    char_indices: std::str::CharIndices<'a>,
}

impl<'a> Reader<'a> {
    pub fn new(source: &'a str, unicode_mode: bool) -> Self {
        Self { source, char_indices: source.char_indices(), unicode_mode, index: 0 }
    }

    pub fn span_position(&self) -> usize {
        let mut char_indices = self.char_indices.clone();

        if self.unicode_mode {
            char_indices.nth(self.index).map_or(self.source.len(), |(i, _)| i)
        } else {
            let mut utf16_units = 0;
            let mut byte_index = 0;
            for (idx, ch) in char_indices {
                if utf16_units == self.index {
                    return idx;
                }

                utf16_units += ch.len_utf16();
                byte_index = idx + ch.len_utf8();
            }
            byte_index
        }
    }

    fn peek_nth(&self, n: usize) -> Option<u32> {
        let nth = self.index + n;

        // ❗ This is not efficient
        if self.unicode_mode {
            self.source.chars().nth(nth).map(|c| c as u32)
        } else {
            #[allow(clippy::cast_lossless)]
            self.source.encode_utf16().nth(nth).map(|u| u as u32)
        }
    }

    pub fn peek(&self) -> Option<u32> {
        self.peek_nth(0)
    }
}
```
</details>

<details>
<summary>Try1: Reuse `Iterator`</summary>

```rs
pub struct Reader<'a> {
    source: &'a str,
    unicode_mode: bool,
    index: usize,
    u8_units: std::str::CharIndices<'a>,
    u16_units: std::str::EncodeUtf16<'a>,
}

impl<'a> Reader<'a> {
    pub fn new(source: &'a str, unicode_mode: bool) -> Self {
        let u8_units = source.char_indices();
        let u16_units = if unicode_mode { "" } else { source }.encode_utf16();
        Self { source, unicode_mode, index: 0, u8_units, u16_units }
    }

    pub fn span_position(&self) -> usize {
        let mut char_indices = self.u8_units.clone();

        if self.unicode_mode {
            char_indices.nth(self.index).map_or(self.source.len(), |(i, _)| i)
        } else {
            let mut utf16_units = 0;
            let mut byte_index = 0;
            for (idx, ch) in char_indices {
                if utf16_units == self.index {
                    return idx;
                }

                utf16_units += ch.len_utf16();
                byte_index = idx + ch.len_utf8();
            }
            byte_index
        }
    }

    fn peek_nth(&self, n: usize) -> Option<u32> {
        let nth = self.index + n;

        if self.unicode_mode {
            self.u8_units.clone().nth(nth).map(|(_, c)| c as u32)
        } else {
            #[allow(clippy::cast_lossless)]
            self.u16_units.clone().nth(nth).map(|u| u as u32)
        }
    }

    pub fn peek(&self) -> Option<u32> {
        self.peek_nth(0)
    }
}
```

</details>

<details>
<summary>Try2: Collect `Vec` at first</summary>
```rs
pub struct Reader<'a> {
    source: &'a str,
    unicode_mode: bool,
    index: usize,
    units: Vec<u32>,
    // NOTE: For now, this exists only for `span_position()` method.
    char_indices: std::str::CharIndices<'a>,
}

impl<'a> Reader<'a> {
    pub fn new(source: &'a str, unicode_mode: bool) -> Self {
        let units = if unicode_mode {
            source.chars().map(|ch| ch as u32).collect::<Vec<_>>()
        } else {
            #[allow(clippy::cast_lossless)]
            source.encode_utf16().map(|cu| cu as u32).collect::<Vec<_>>()
        };

        Self { source, unicode_mode, index: 0, units, char_indices: source.char_indices() }
    }

    pub fn span_position(&self) -> usize {
        let mut char_indices = self.char_indices.clone();

        if self.unicode_mode {
            char_indices.nth(self.index).map_or(self.source.len(), |(i, _)| i)
        } else {
            let mut utf16_units = 0;
            let mut byte_index = 0;
            for (idx, ch) in char_indices {
                if utf16_units == self.index {
                    return idx;
                }

                utf16_units += ch.len_utf16();
                byte_index = idx + ch.len_utf8();
            }
            byte_index
        }
    }

    fn peek_nth(&self, n: usize) -> Option<u32> {
        let nth = self.index + n;
        self.units.get(nth).copied()
    }

    pub fn peek(&self) -> Option<u32> {
        self.peek_nth(0)
    }
}
```

</details>

In my environment, `cargo bench --bench parser` said Try2 is faster.

I'll seek for Try3 for a while.

leaysgur commented:
@Boshen ^ How do you think? (maybe also related to https://github.com/oxc-project/oxc/pull/4242#discussion_r1705694511)

Sysix commented:
Hello @leaysgur 

for `eslint/no-useless-backreference` it would be really nice to have the AST-Span of the Backreference.
But there is a difference with RegexLiterals and the RegExp Obj: The Backlash is double escaped. 

See:
```
  ⚠ no back reference
   ╭─[no_useless_backreference.tsx:1:6]
 1 │ /(b)(\2a)/
   ·      ──
   ╰────
vs.
  ⚠ no back reference
   ╭─[no_useless_backreference.tsx:1:6]
 1 │ new RegExp("(b)(\\2a)");
   ·                 ───
   ╰────
```

Did you considered this use case for escaped backslashes?
Do we want to catch the doubled backslash?

leaysgur commented:
@Sysix Thanks for your comment!

> nice to have the AST-Span of the Backreference.

The current AST for backref already holds `span`.
Please check https://github.com/oxc-project/oxc/pull/3824/files#diff-8a24d853ad0d7af27dc285d21a4b1f2d6aa97d2fab6153fa441805750266d7e4R245-R248 👀 

> Did you considered this use case for escaped backslashes?

Yes, but as a RegExp parser, I do not specifically address backslash escaping (rather escape sequences).
Because it's a topic of string literals themselves, not just RegExp.

For now, the treatment of pattern `\\2` is (escaped)backslash and (number)2.

My understanding may be wrong and I'm not sure how OXC parser handle these escapes. 😅

Sysix commented:
> Nope… For this reason, we may need to implement a lexer layer...? (And also isolate the surrogate pair issue?)

I solved it in my PR with a simple bool and counted 2 up instead of 1 for the span offset (start) :) 
The PR is closed because I am waiting for your PR 🤩 

https://github.com/oxc-project/oxc/pull/4096/files#diff-afa407d7754315f9d88597efc36c532c89aabeb1e8639433a3401252efd5b92dR111-R115

Boshen reviewed (approved):
This is art.

leaysgur committed:
feat(regex_parser): Implement `RegExp` parser (#3824)

Part of #1164

## Progress updates 🗞️

Waiting for the review and advice, while thinking how to handle escaped string when `new RegExp(pat)`.

## TODOs

- [x] `RegExp(Literal = Body + Flags)#parse()` structure
- [x] Base `Reader` impl to handle both unicode(u32) and utf-16(u16) units
- [x] Global `Span` and local offset conversion
- [x] Design AST shapes
  - [x] Keep `enum` size small by `Box<'a, T>`
  - [x] Rework AST shapes
- [x] Split body and flags w/ validating literal
- [x] Parse `RegExpFlags`
- [x] Parse `RegExpBody` = `Pattern`
- [x] Parse `Pattern` > `Disjunction`
- [x] Parse `Disjunction` > `Alternative`
- [x] Parse `Alternative` > `Term`
- [x] Parse `Term` > `Assertion`
	- [x] Parse `BoundaryAssertion`
	- [x] Parse `LookaroundAssertion`
- [x] Parse `Term` > `Quantifier`
- [x] Parse `Term` > `Atom`
	- [x] Parse `Atom` > `PatternCharacter`
	- [x] Parse `Atom` > `.`
	- [x] Parse `Atom` > `\AtomEscape`
		- [x] Parse `\AtomEscape` > `DecimalEscape`
		- [x] Parse `\AtomEscape` > `CharacterClassEscape`
			- [x] Parse `CharacterClassEscape` > `\d, \D, \s, \S, \w, \W`
			- [x] Parse `CharacterClassEscape` > `\p{UnicodePropertyValueExpression}, \P{UnicodePropertyValueExpression}`
		- [x] Parse `\AtomEscape` > `CharacterEscape`
			- [x] Parse `CharacterEscape` > `ControlEscape`
			- [x] Parse `CharacterEscape` > `c AsciiLetter`
			- [x] Parse `CharacterEscape` > `0`
			- [x] Parse `CharacterEscape` > `HexEscapeSequence`
			- [x] Parse `CharacterEscape` > `RegExpUnicodeEscapeSequence`
			- [x] Parse `CharacterEscape` > `IdentityEscape`
		- [x] Parse `\AtomEscape` > `kGroupName`
	- [x] Parse `Atom` > `[CharacterClass]`
    	- [x] Parse `[CharacterClass]` > `ClassContents` > `[~UnicodeSetsMode] NonemptyClassRanges`
    	- [x] Parse `[CharacterClass]` > `ClassContents` > `[+UnicodeSetsMode] ClassSetExpression`
          - [x] Parse `ClassSetExpression` > `ClassUnion`
          - [x] Parse `ClassSetExpression` > `ClassIntersection`
          - [x] Parse `ClassSetExpression` > `ClassSubtraction`
          - [x] Parse `ClassSetExpression` > `ClassSetOperand`
          - [x] Parse `ClassSetExpression` > `ClassSetRange`
          - [x] Parse `ClassSetExpression` > `ClassSetCharacter`
	- [x] Parse `Atom` > `(GroupSpecifier)`
	- [x] Parse `Atom` > `(?:Disjunction)`
- [x] Annex B
    - [x] Parse `QuantifiableAssertion`
	- [x] Parse `ExtendedAtom`
      - [x] Parse `ExtendedAtom` > `\ [lookahead = c]`
      - [x] Parse `ExtendedAtom` > `InvalidBracedQuantifier`
      - [x] Parse `ExtendedAtom` > `ExtendedPatternCharacter`
      - [x] Parse `ExtendedAtom` > `\AtomEscape` > `CharacterEscape` > `LegacyOctalEscapeSequence`
- [x] Early errors
	- [x] Pattern :: Disjunction(1/2)
	- [x] Pattern :: Disjunction(2/2)
	- [x] QuantifierPrefix :: { DecimalDigits , DecimalDigits }
	- [x] ExtendedAtom :: InvalidBracedQuantifier (Annex B)
	- [x] AtomEscape :: k GroupName
	- [x] AtomEscape :: DecimalEscape
	- [x] NonemptyClassRanges :: ClassAtom - ClassAtom ClassContents(1/2)
	- [x] NonemptyClassRanges :: ClassAtom - ClassAtom ClassContents(2/2)
	- [x] NonemptyClassRanges :: ClassAtom - ClassAtom ClassContents(Annex B)
	- [x] NonemptyClassRangesNoDash :: ClassAtomNoDash - ClassAtom ClassContents(1/2)
	- [x] NonemptyClassRangesNoDash :: ClassAtomNoDash - ClassAtom ClassContents(2/2)
	- [x] NonemptyClassRangesNoDash :: ClassAtomNoDash - ClassAtom ClassContents(Annex B)
	- [x] RegExpIdentifierStart :: \ RegExpUnicodeEscapeSequence
	- [x] RegExpIdentifierStart :: UnicodeLeadSurrogate UnicodeTrailSurrogate
	- [x] RegExpIdentifierPart :: \ RegExpUnicodeEscapeSequence
	- [x] RegExpIdentifierPart :: UnicodeLeadSurrogate UnicodeTrailSurrogate
	- [x] UnicodePropertyValueExpression :: UnicodePropertyName = UnicodePropertyValue(1/2)
	- [x] UnicodePropertyValueExpression :: UnicodePropertyName = UnicodePropertyValue(2/2)
	- [x] UnicodePropertyValueExpression :: LoneUnicodePropertyNameOrValue(1/2)
	- [x] UnicodePropertyValueExpression :: LoneUnicodePropertyNameOrValue(2/2)
	- [x] CharacterClassEscape :: P{ UnicodePropertyValueExpression }
	- [x] CharacterClass :: [^ ClassContents ]
	- [x] NestedClass :: [^ ClassContents ]
	- [x] ClassSetRange :: ClassSetCharacter - ClassSetCharacter
- [x] Add `Span` to `Err(OxcDiagnostic::error())` calls
- [x] Perf improvement
	- [x] `Reader#peek()` should avoid `iter.next()` equivalent
	- [x] ~~Use `char` everywhere and split and push 2 surrogates(pair) for `Character`?~~
	- [x] ~~Try 1(+1) loop parsing for capturing groups?~~

## Follow up

- [x] @Boshen Test suite > #4242
  - [x] Investigate CI errors...
- Next...
  - Support ES2025 Duplicate named capturing groups?
  - Support ES20XX Stage3 Modifiers?

leaysgur commented:
@Sysix Sorry to bother you from already closed PR.

But I finally found that we do not need to care about escaped backslash issue you mentioned.

Please see `example/parse_file` in https://github.com/oxc-project/oxc/pull/5106
